Fixed #6978: DI Container is not reset when destroying application in functional tests

ivokund opened the pull request:
Added code to reset the DI container when destroying the application

Commit:
Fixed #6978: Dependency Injection Container not reset when destroying application in functional tests

samdark commented:
@yiisoft/core-developers, @Ragazzo any issues with the change?

klimov-paul commented:
I am unsure: it could be someone relies on DI for test environment setup.
It could be better to save original instance if `Container` before running the test and then restore it, rather then simply recreating it.

samdark commented:
@klimov-paul environment should be re-created for each test case i.e. it should be isolated.

Ragazzo commented:
i dont think that it will affect anything , since tests should be done in isolation , however we should recheck that DI - container is recreated each time in mockApplication()

cebe commented on framework/CHANGELOG.md:
```diff
@@ -6,6 +6,7 @@ Yii Framework 2 Change Log
 
 - Bug #6919: Fixed wrong namespaces under advanced application's TestCase classes (ivokund)
 - Bug #6969: `yii\helpers\ArrayHelper::htmlEncode()` and `htmlDecode()` should not remove non-string data (qiangxue)
+- Bug #6978: DI Container is not reset when destroying application in functional tests (ivokund)
```
changelog has to be in https://github.com/yiisoft/yii2/blob/master/extensions/codeception/CHANGELOG.md

cebe commented:
> however we should recheck that DI - container is recreated each time in mockApplication()

:+1:

ivokund commented:
Okay, do I understand this correctly now?
To be precise, we should do `Yii::$container = null` (to destroy is to set to default value) in `destroyApplication()` and  `Yii::$container = new Container` in `mockApplication()` (to mimick the first load of `Yii` class?)
```php
class Yii extends \yii\BaseYii
{
}

spl_autoload_register(['Yii', 'autoload'], true, true);
Yii::$classMap = include(__DIR__ . '/classes.php');
Yii::$container = new yii\di\Container;
```

ivokund commented:
Ok, setting it to null is not a good idea, because `yii\base\Application::__construct` needs the container to be set already.

Commit:
Fixed #6978: Dependency Injection Container not reset when destroying application in functional tests

Added the change to the correct changelog, added the reset of DI container also to mockApplication

ivokund commented:
Added the change to the correct changelog, added the reset of DI container also to `mockApplication`.

Did you mean something like this by "recheck"?

samdark commented:
Merged. Thanks! https://github.com/yiisoft/yii2/commit/a6dd9d37d5c5b7e5fac4edeb89c35cbc961df369
